### PR TITLE
adding three-phase simulator including water evaporation and salt precipitation

### PR DESCRIPTION
**Pull request by plgbrts**
The main contribution of this PR:

-  a three-phase simulator including water evaporation and salt precipitation
- calculation of evaporated water production plus correction of salt production rate (evaporated water does not contain salt)

This PR depends on 
- https://github.com/OPM/opm-common/pull/2989
- https://github.com/OPM/opm-material/pull/497
- https://github.com/OPM/opm-models/pull/688

### CODE REVIEW

**Comment by plgbrts** [comment]
jenkins build this opm-common=2989 opm-material=497 opm-models=688 please

**Comment by plgbrts** [comment]
jenkins build this opm-common=2989 opm-material=497 opm-models=688 please

**Comment by plgbrts** [comment]
There is one test case (`SPE1CASE1_PRECSALT`) failing but that is to expected because of a code modification  (related to updating the porosity). The code is ready for review.

File: opm/simulators/wells/MultisegmentWellEval.cpp [hunk 1]
```diff
@@ -692,9 +693,9 @@ computeSegmentFluidProperties(const EvalWell& temperature,
                         rv = rvmax;
                     }
                     b[gasCompIdx] =
-                        FluidSystem::gasPvt().inverseFormationVolumeFactor(pvt_region_index, temperature, seg_pressure, rv);
+                        FluidSystem::gasPvt().inverseFormationVolumeFactor(pvt_region_index, temperature, seg_pressure, rv, rvw);
                     visc[gasCompIdx] =
-                        FluidSystem::gasPvt().viscosity(pvt_region_index, temperature, seg_pressure, rv);
+                        FluidSystem::gasPvt().viscosity(pvt_region_index, temperature, seg_pressure, rv, rvw);
                     phase_densities[gasCompIdx] = b[gasCompIdx] * surf_dens[gasCompIdx]
                                                 + rv * b[gasCompIdx] * surf_dens[oilCompIdx];
```
[the remark this answers is not in the document]
Comment (totto82) [reply opm/simulators/wells/MultisegmentWellEval.cpp:700]: And you also need to modify it in the standard well model here 

https://github.com/OPM/opm-simulators/blob/cd663351f6fe4f0ada7902421b7c1e475e9631cf/opm/simulators/wells/StandardWellEval.cpp#L943

**Review by totto82 — CHANGES_REQUESTED** [review]
Thanks for the contribution. The density needs adaption due to rvw. b

**Comment by plgbrts** [comment]
jenkins build this opm-common=2989 opm-material=497 opm-models=688 please

**Comment by plgbrts** [comment]
jenkins build this opm-common=2989 opm-material=497 opm-models=688 please

**Comment by plgbrts** [comment]
jenkins build this opm-material=497 opm-models=688 please

**Comment by plgbrts** [comment]
jenkins build this opm-material=497 opm-models=688 please

**Comment by plgbrts** [comment]
jenkins build this opm-material=497 opm-models=688 please

**Comment by plgbrts** [comment]
Thanks for the review and pointing out the issues! The gas property calculations to include water vapor (rvw) dependency in the standard well code are fixed.
A similar correction in the multisegment well (`MSW`) code I started but is more involved and gave some MSW test failures. I propose to make a separate (WIP) PR for that change. 
The other rvw dependency missing you indicated is in `ebos/equil/initstateequil.hh` and is related to equilibration initialization.    Equilibration initialization to include also water vapor is on our wishlist but not implemented yet ( it will  require implementation of a `RVWVD` table similar to `RVVD`) .  So far we use explicit initialization for our water vapor testcases.
The current PR is thus for water evaporation and, for now, assuming explicit initialization and standard wells ( no `MSW`). The extensions for equilibration initialization and `MSW` I propose to do via other PRs.
The single test case (SPE1CASE1_PRECSALT) failing  is to expected because of a code modification as mentioned earlier.

**Comment by totto82** [comment]
benchmark opm-material=497 opm-models=688 please

**Comment by totto82** [comment]
> A similar correction in the multisegment well (MSW) code I started but is more involved and gave some MSW test failures. I propose to make a separate (WIP) PR for that change.

Ok. Please add a throw if water evaporation and msw is both is true. (I noticed we currently don't support Brine + MSW as well) 
https://github.com/OPM/opm-simulators/blob/ea82bebb3a21afa7960a737f2e8b4f4e3b7b06a8/opm/simulators/wells/MultisegmentWell_impl.hpp#L55

**Comment by totto82** [comment]
I support merging this. I just want to assure we don't have a unexpected performance hit. Lets see what benchmark says.

**Comment by plgbrts** [comment]
jenkins build this opm-material=497 opm-models=688 please

**Comment by totto82** [comment]
> benchmark opm-material=497 opm-models=688 please


@akva2 is this possible? I really want to double check that this does not impact performance on the blackoil cases.

**Comment by akva2** [comment]
uhm, i'm not responsible for benchmark scripts so i don't definitely know but i believe it should be. @alfbr ?

**Comment by alfbr** [comment]
> @akva2 is this possible? I really want to double check that this does not impact performance on the blackoil cases.

Yes, it is supported to add pull requests. Just as with Jenkins. I do think we got the syntax in a slightly different order, which we probably should fix. Let me try.

**Comment by alfbr** [comment]
benchmark please opm-material=497 opm-models=688

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test | Configuration | Relative |
| ---- | ------------- | -------- |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=1644

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test | Configuration | Relative |
| ---- | ------------- | -------- |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=1644

**Comment by alfbr** [comment]
It seems both syntax worked, but none seemed to run.

**Comment by alfbr** [comment]
Just checked, only the first pull request seems to be pulled in. Will try to fix.

**Comment by alfbr** [comment]
benchmark please opm-material=497 opm-models=688

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                                     | Relative |
| ------- | ------------------------------------------------- | -------- |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1        | 1.009    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8        | 1.006    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1        | 0.997    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8        | 0.994    |
| opm-git | OPM Benchmark: flow_mpi_norne_4c_msw - Threads: 1 | 0.998    |
| opm-git | OPM Benchmark: flow_mpi_norne_4c_msw - Threads: 8 | 0.977    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=1645

**Comment by totto82** [comment]
> Benchmark result overview:

The run-times are fine. I will start the update of data process for the one test that naturally fails.

**Comment by totto82** [comment]
jenkins build this opm-material=497 opm-models=688 update_data please

**Comment by totto82** [comment]
jenkins build this opm-material=497 opm-models=688 opm-tests=760 please

**Comment by totto82** [comment]
jenkins build this opm-material=497 opm-models=688 opm-tests=760 please

**Comment by totto82** [comment]
jenkins build this opm-material=497 opm-models=688 please

**Comment by totto82** [comment]
jenkins build this opm-material=497 opm-models=688 update_data please

**Comment by totto82** [comment]
jenkins build this opm-material=497 opm-models=688 opm-tests=760 please

**Comment by bska** [comment]
> opm-tests=760

That's guaranteed to fail.  I had to close that PR due to merge conflicts with OPM/opm-tests#763 (generated from OPM/opm-common#3021).  Also, please note that you should **generally** run build checks of multi-repo PR chains from the most upstream PR&mdash;in this case OPM/opm-material#497.  The reason is that we, meaning the Jenkins CI system, don't really run unit tests in upstream repositories.

**Comment by bska** [comment]
Merging this now since the upstream PRs are all in the build queue.

**Pull request merged**